### ISSUE TITLE
Instance resolution

### ISSUE THREAD

**Issue by truizlop** [arrow-kt/kotlin#5]
When an instance needs to be resolved, where should the compiler look for it? In the general case, we would have:

**Typeclass definition**
```kotlin
package A
interface Typeclass<T>
```

**Instance definition**
```kotlin
package B
extension class TypeclassInstance : Typeclass<ConcreteType>
```

**Function definition**
```kotlin
package C
fun <A> f(a : RegularArgument, with instance : Typeclass<A>)
```

**Function call**
```kotlin
package D
f(value) // instance is implicitly created and passed as an argument
```

## Proposal for resolution order

1. Look for instance in the scope of the function.
```kotlin
fun <A> g(with g_instance : Typeclass<A>) {
  f(value) // instance resolved to g_instance
}
```

2. Look for an `extension class` implementing `Typeclass<A>` in the `import` section of the file where `f` is defined.

3. Look for an `extension class` implementing `Typeclass<A>` in the package where the function `f` is being called (`package D`).

4. Look for an `extension class` implementing `Typeclass<A>` in the `companion object` for `A`.

5. Look for an `extension class` implementing `Typeclass<A>` in the `companion object` for `Typeclass`.

6. Compilation error: Unable to resolve implicit instance.

**Comment by JorgeCastilloPrz** [comment]
Looks good overall, you're doing a great job, really. Some doubts I have:
* is `2` looking for explicit imports of instances for `Typeclass<A>`? what would happen if somebody imports a bunch of them together using wildcard?
* also for `2`, shouldn't we search for it in the `import` section where `g` is defined? (the enclosing function, the caller) 🤔 not sure though
* for `4`, what if `A` is `Int` for example? Do all the boxed types support companions already?

**Comment by truizlop** [comment]
- You can use both explicit and wildcard imports. If you use explicit, we check the explicit class that you imported. If you use wildcard, we examine the whole package contents looking for classes that are potential candidates for the typeclass.

- You are right, I think I made a mistake. We need to look in the import section of those places where `f` is being called, not defined.

- Same as above. We cannot resolve instances on `f` definition, only when it is called (when we have enough information of the concrete types we need to look for an instance of the typeclass for such types).

- If `A` does not have a companion, we skip it. In the case of Boxed types, or types that you don't own, you can still define instances in a) the companion of the typeclass, or b) in a separate class and import it. In both cases it will be resolved.

**Comment by truizlop** [comment]
@pakoito Ok so that means that I cannot have something like:

```kotlin
extension IntSumSemigroup : Semigroup<Int> {
  override fun Int.combine(b : Int) : Int = this + b
}

extension IntProductSemigroup : Semigroup<Int> {
  override fun Int.combine(b : Int) : Int = this * b
}
```

Only one of them can be marked as `extension`, right? Is it something that we need to enforce checking? How would the resolution process be, assuming coherence is implemented like this?

**Comment by pakoito** [comment]
Yes, that's what coherence means. We'll need to enforce checking it ("Exception: overlapping instances in files X and Y"). And the resolution is just against a global map.

>So you'll have to traverse the whole codebase and get a map of instances, their location, and their signature including generic parameters. When they're requested implicitly, just pick it up from the map, check that the signature matches correctly, and plug it in directly.

**Comment by truizlop** [comment]
What happens then when we have something like this?

```kotlin
fun <A> plus(a : A, b : A, with semigroup : Semigroup<A>) : A = a.combine(b)
fun <A> dup(a : A) : A = plus(a, a)
```

In this case, the invocation of `plus(a, a)` cannot be resolved since we don't have information about `A` when function is called. It is necessary to add:

```kotlin
fun <A> dup(a : A, with semigroup : Semigroup<A>) : A = plus(a, a)
```

Then, `semigroup` for `plus(a, a)` is the one resolved for `dup`, because likely there won't be anything on the map matching `Semigroup<A>`.

A similar thing may happen when resolving nested instances, like:

```kotlin
extension class OptionSemigroup<A>(with val semigroup : Semigroup<A>) : Semigroup<Option<A>> {
  // implementation
}

fun maybePlus(a : Option<Int>, b : Option<Int>, with semigroup : Semigroup<Option<Int>>) : Option<Int> =
  a.combine(b)
```

There may not be something with `Semigroup<Option<Int>>` in the map, but we'd need to resolve both the `OptionSemigroup` and the `IntSemigroup`, and combine them to provide the instance. I think this is not trivial and I still don't know exactly how to approach it.

Other hard cases:

```kotlin
class Foo
class Bar : Foo

extension class FooSemigroup : Semigroup<Foo> { ... }

fun f(a : Foo, b : Foo, with semigroup : Semigroup<Foo>) : Foo = a.combine(b)

val x = f(Bar(), Bar())
```

From this call, we would start looking for a `Semigroup<Bar>`, which is not present in the map, but a `Semigroup<Foo>` is valid to resolve this call.

**Comment by ClaireNeveu** [comment]
@truizlop I don't understand your examples. In the case of
```
extension class OptionSemigroup<A>(with val semigroup : Semigroup<A>) : Semigroup<Option<A>> {
  // implementation
}

fun maybePlus(a : Option<Int>, b : Option<Int>, with semigroup : Semigroup<Option<Int>>) : Option<Int> =
  a.combine(b)
```
_only_ `OptionSemigroup` needs to be resolved, `IntSemigroup` does not match the constraint here and we don't need to "combine" the two, we just supply `OptionSemigroup`.

An actual example of nested type-classes is
```
extension class IdentityMonoid<A>(with m: Monoid<A>): Monoid<Identity<A>> {
   fun empty(): Identity<A> = Identity(m.empty())
   fun append(a: Identity<A>, b: Identity<A>): Identity<A> = Identity(a.run().append(b.run()))
}

fun foo(i: Identity<Int>, j: Identity<Int>, with Monoid<Identity<Int>>)
```
In this case it's pretty simple to see how this resolution would work.
1. Look for an instance that matches `Monoid<Identity<Int>>`; `IdentityMonoid` matches this constraint.
2. See if the matching instance has any constraints; `IdentityMonoid` does have constraints.
3. Begin a new search to fulfill _those_ constraints.
3a. If the search succeeds, this instance matches.
3b. If the search fails, this instance does not match, continue searching for instances that match `Monoid<Identity<Int>>`

Then for the runtime portion, `IdentityMonoid` is just a class that takes another class as an argument. All instances get their constraints passed to them as arguments (or inlined into them if we end up doing any specialization).

Your other example of having `IntSemigroup` and `IntMonoid : IntSemigroup` would be an error since those are duplicate instances. When type-classes have superclass constraints, you should only implement the type-class at the bottom of the type-hierarchy.

As far as instances not defined in a companion (orphan instances) go, I laid out some ideas for that in my original coherence proposal on the KEEP. My recommendation is to disallow orphans now and revisit this later with the goal of tackling it once Java 9 modules land (stabilizing the behavior before then might complicate things). Haskell allows orphans but that drastically complicates the resolution mechanism (you have to scan the whole codebase, very unmodular). Rust completely forbids them. Scala—because its type-classes are faked—allows them but its resolution mechanism is notoriously confusing and unreliable.

**Comment by ClaireNeveu** [comment]
For the record, I'd be happy to contribute code toward the resolution implementation but I'm not familiar with the codebase so I'm not sure where to start (I'm also quite busy until the end of the year which hasn't afforded me the time to do a solo deep-dive).

**Comment by truizlop** [comment]
@raulraja what you mention is in conflict with what @pakoito stated: coherence = one instance per typeclass and type, per codebase. In the example there are two instances of Monoid<Int> and of course there could be a disambiguation problem. But if we check which one is imported, then they could in fact be disambiguated, right? I think the problem in this case is that, as far as I have investigated, the import list is not easily accessible (at least I haven't been able to retrieve it).

@ClaireNeveu in my example about `OptionSemigroup` I mentioned that I need an `IntSemigroup` because it is needed in order to instantiate an `OptionSemigroup`. In the following code:

```kotlin
val result = plus(Some(1), Some(2))
```

We will start looking for something that implements `Semigroup<Option<Int>>`. We could match with `OptionSemigroup`, which implements `Semigroup<Option<A>>`, and assume `A = Int`. Then we try to instantiate `OptionSemigroup`, but we realize it needs another implicit parameter, so we need to resolve it looking for `Semigroup<A>`. Since `A = Int`, we need to look for `Semigroup<Int>`. We would need to proceed recursively in this case until we manage to obtain all instances, or fail. As you mentioned, the moment we add subtyping to the equation, it complicates things enormously; I'd suggest leaving it off the table for now.

The status of the implementation, right now: I am able to resolve instances which do not have nested instances, looking for them in (1) parameters of the function where it is called; (2) package where the function call is defined; (3) type companion; and (4) typeclass companion. So far it works nicely, but it does not cover the case @raulraja mentioned for `arrow-effects-coroutines`. I am not exactly sure how to approach that case without scanning the whole codebase (which may not even be possible). Any hints about this would be very appreciated.

**Comment by raulraja** [comment]
@truizlop could the scan be performed in all subpackages under the type class or data type declared package? Most interesting instances for Tagless final like Effect for Deferred, Observable, Flux etc face this problem and we cant just have one library depend on all of those third party libraries at once.

**Comment by truizlop** [comment]
@raulraja could you provide an example?

**Comment by raulraja** [comment]
module: `arrow-effects`
```kotlin
package arrow.effects
interface Effect<F>
```
module: `arrow-effects-coroutines`
```kotlin
package arrow.effects.coroutines
extension class DefferedEffect: Effect<ForDeferredK>
```

In this case we can't make the base type class module `arrow-effects` depend on the `kotlin-coroutines` but the actual `arrow-effects-coroutines` module can establish such dependency.

We can't have `arrow-effects` depend on further modules because otherwise users would get transitive dependencies not just to coroutines but also of rx2 etc.. and any other integration we add.

If we scan sub-packages for instances we allow this relationship. While this may not be initially coherent it would allow submodules to declare instances which are resolved simply scanning subpackages of `arrow.effects` which is where the type class is defined. Note this is not currently an issue because we define `DefferedK` ourselves but in the future once we have support for Kinds this will be a problem because `Effect<Deferred>` could be define directly over the third party data type.

**Comment by pakoito** [comment]
Rust has this same limitation too apparently, for the same reasons maybe.
